[BUGFIX beta] Deprecate passing non-Array enumerables to addObjects() and pushObjects()

bendemboski opened the pull request:
This is the result of the discussion in #9843 

pushObjects(), unshiftObjects(), removeObjects() and setObjects() were all documented to accept enumerables, but would only function properly when passed a native array, so update their documentation.

pushObjects() has an assert ensuring that the argument is an enumerable or array-like object. It should really be asserting that the argument is a native array, but changing it would cause calls passing in an enumerable that now produce incorrect behavior to start throwing an exception, so instead just add a deprecation warning in case the argument is not a native array.

addObjects() was documented to accept enumerables, and would behave correctly when passed an enumerable, which is not consistent with its sister methods and therefore not desired behavior. So update the documentation and add a deprecation warning in case the argument is not a native array.

Also take the opportunity to implement some addObjects() unit tests.

Commit:
[BUGFIX beta] Deprecate passing non-Array enumerables to addObjects() and pushObjects()

pushObjects(), unshiftObjects(), removeObjects() and setObjects() were all documented to accept enumerables, but would only function properly when passed a native array, so update their documentation.

pushObjects() has an assert ensuring that the argument is an enumerable or array-like object. It should really be asserting that the argument is a native array, but changing it would cause calls passing in an enumerable that now produce incorrect behavior to start throwing an exception, so instead just add a deprecation warning in case the argument is not a native array.

addObjects() was documented to accept enumerables, and would behave correctly when passed an enumerable, which is not consistent with its sister methods and therefore not desired behavior. So update the documentation and add a deprecation warning in case the argument is not a native array.

Also take the opportunity to implement some addObjects() unit tests.

wycats commented:
I think there are a few things going on here:

1) We are deprecating `Ember.Enumerable`, so we should definitely not add new support for `Ember.Enumberable`, and deprecate anything that relies on it. We should revisit this question in general once we align with the ES6 iterable protocol.
2) We are not deprecating `Ember.Array` (and `ArrayProxy` is an `Ember.Array`) so it seems like we should add broader support.
3) The `Object`/`Objects`-suffixed methods are "souped up" Ember versions of methods, so they should probably all support `Ember.Array` (but not `Ember.Enumerable`).

stefanpenner commented:
I think we actually will only be supporting Arrays, and a potential future of @@iterables. That being said, this PR seems to have gone stale, and @wycats feedback from 8 months ago hasn't been applied.

I suspect with some adjustments we would consider this, so for anyone interested feel free to submit a PR with the above feedback implemented.

bendemboski commented:
@stefanpenner sorry, I guess I didn't realize that there was feedback in @wycats comment that needed to be addressed. I'm still not sure I see what that would be.

This PR cleans up documentation and adds deprecation warnings to move away from the usage of ```Ember.Enumerable```s in favor of already-supported native arrays. It does not add support for ```Ember.Array```s -- is that the unaddressed feedback? That you don't want to take a PR that fixes the documentation and deprecates ```Ember.Enumerable```s without also adding the broader support for ```Ember.Array```s? Or is there something else that I'm missing here?
